### ISSUE TITLE
vimPlugins.windows-nvim: dependency not found

### ISSUE THREAD

**Issue by ozkutuk** [NixOS/nixpkgs#312050]
### Describe the bug
One of the dependencies of the plugin, `middleclass`, is not found by the plugin:
```
E5113: Error while calling lua chunk: ...ovimPackages/start/animation.nvim/lua/animation/init.lua:1: module 'middleclass' not found:                                                           
        no field package.preload['middleclass']                                                                                                                                                
        no file '/nix/store/kf5k5qa53g9rrsch256vqgnr7al080al-luajit-2.1.1693350652-env/share/lua/5.1/middleclass.lua'                                                                          
        no file '/nix/store/kf5k5qa53g9rrsch256vqgnr7al080al-luajit-2.1.1693350652-env/share/lua/5.1/middleclass/init.lua'                                                                     
        no file '/nix/store/kf5k5qa53g9rrsch256vqgnr7al080al-luajit-2.1.1693350652-env/lib/lua/5.1/middleclass.so'                                                                             
stack traceback:                                                                                                                                                                               
        [C]: in function 'require'                                                                                                                                                             
        ...ovimPackages/start/animation.nvim/lua/animation/init.lua:1: in main chunk                                                                                                           
        [C]: in function 'require'                                                                                                                                                             
        ...windows.nvim/lua/windows/lib/resize-windows-animated.lua:5: in main chunk                                                                                                           
        [C]: in function 'require'                                                                                                                                                             
        ...myNeovimPackages/start/windows.nvim/lua/windows/init.lua:8: in function 'setup'                                                                                                     
        /home/ozkutuk/.config/nvim/lua/extra.lua:297: in main chunk                                                                                                                            
        [C]: in function 'require'                                                                                                                                                             
        /home/ozkutuk/.config/nvim/init.lua:5: in main chunk
```
It is interesting, because the dependency is actually listed in the overrides:
```nix
  windows-nvim = super.windows-nvim.overrideAttrs {
    dependencies = with self; [ luaPackages.middleclass animation-nvim ];
  };
```

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.7, NixOS, 24.05 (Uakari), 24.05.20240419.5c24cf2`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.21.2`
 - nixpkgs: `/etc/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by teto** [comment]
should be solved by the PR linked in https://github.com/NixOS/nixpkgs/issues/318925 . Will take a few days before reaching unstable.

**Comment by teto** [comment]
can you check if this still happens on unstable ?

**Comment by gepbird** [comment]
> can you check if this still happens on unstable ?

@teto yes, putting `windows-nvim` to nvim plugins and running `nvim -c "=require('windows').setup()"` will fail with the same error message: `E5108: Error executing lua ...ovimPackages/start/animation.nvim/lua/animation/init.lua:1: module 'middleclass' not found:`. See https://github.com/NixOS/nixpkgs/pull/312311#issuecomment-2220371961 for explanation.

**Comment by teto** [comment]
The way it is implemented in nixpkgs is that when the neovim wrapper creates the packpath directory with the plugins, it looks at the propagedBuildInputs and puts the lua dependencies into the LUA_PATH wrapped with neovim.


```
  windows-nvim = super.windows-nvim.overrideAttrs {
    dependencies = with self; [ luaPackages.middleclass animation-nvim ];
  };
```
wont work because it will copy middleclass files in the packpath directory with the prepended lua files. Either you convert `luaPackages.middleclass` to a neovim plugin (it takes the lua package and removes the lua folder prefixes) or you make windows.nvim a luarocks package. Uploading a windows.nvim rockspec to luarocks.org has the added benefit that nixpkgs can automatically deduce its dependencies, same with rocks.nvim and more recently lazy.nvim. 
It's quite easy to add a package to luarocks.org via NURR see https://github.com/nvim-neorocks/rocks.nvim/wiki/Plugin-support

**Comment by gepbird** [comment]
> Either you convert luaPackages.middleclass to a neovim plugin...

@teto thanks this worked well, I updated #312311.

NURR seems interesting, I will look into it later. If this will be the preferred way of packaging neovim plugins with dependencies, should we start migrating plugins to luarocks?

**Comment by teto** [comment]
if you look at maintainers/scripts/luarocks-packages.csv, you will see there are already several neovim plugins packaged as luarocks plugins. And the approach seems to gain traction. I've edited the landing page of https://github.com/nvim-neorocks/ to explain what we want to achieve at neorocks. 

Adding the plugins to nurr is as easy as https://github.com/nvim-neorocks/nurr/pull/14/files It would be best if the plugin author would take that task upon themselves via https://github.com/nvim-neorocks/luarocks-tag-release/ as described in my previous comment's link but that might be too early (do you think winmdows-nvim author would welcome the idea to put his plugin on luarocks ?).
Then the luarocks packages can be added to maintainers/scripts/luarocks-packages.csv and the luarocks package set updated as per the documentation.
Then as you did in your PR, you can add in vim/overrides.nix `windows-nvim = buildNeovimPlugin { luaAttr = ""; }`. The nixpkgs side of things could be made less redundant I admit.

**Comment by gepbird** [comment]
@teto thanks for all of the info, it was pretty helpful. I published windows-nvim (and a few other plugins) to neorocks.

> do you think winmdows-nvim author would welcome the idea to put his plugin on luarocks ?

I don't think so, they were very inactive in the project recently, and slightly active on GitHub.

> if you look at maintainers/scripts/luarocks-packages.csv, you will see there are already several neovim plugins packaged as luarocks plugins. And the approach seems to gain traction. I've edited the landing page of https://github.com/nvim-neorocks/ to explain what we want to achieve at neorocks.

I took a look at a few plugins in the luarocks csv, namely `gitsigns.nvim`, `nvim-cmp`, `rustaceanvim` and all of those are also in the generated vim plugin list. Are these luarock packages used anywhere?

> Then the luarocks packages can be added to maintainers/scripts/luarocks-packages.csv and the luarocks package set updated as per the documentation.
Then as you did in your PR, you can add in vim/overrides.nix windows-nvim = buildNeovimPlugin { luaAttr = ""; }. The nixpkgs side of things could be made less redundant I admit.

I'll do this soon.

**Comment by teto** [comment]
> I took a look at a few plugins in the luarocks csv, namely gitsigns.nvim, nvim-cmp, rustaceanvim and all of those are also in the generated vim plugin list. Are these luarock packages used anywhere?

They exist just to be converted into neovim plugins via `buildNeovimPlugin`. `buildNeovimPlugin` just installs the lua package with a different luarocks config that works natively with neovim.

There has been a lot of improvements on the neovim and lua side in the last nixos release (and in the neovim plugin ecosystem in part thanks to neorocks) and I think we can do much better in terms of UX on the nix side. I've had ideas for years but this takes time :'(

**Comment by gepbird** [comment]
@teto

I'm having some trouble with adding it as a lua module. There's animation.nvim [on luarocks](https://luarocks.org/modules/neorocks/animation.nvim), however using any of these for luarocks-packages.csv makes the updater crash:
```csv
name,rockspec,ref,server,version,luaversion,maintainers
animation.nvim,,,,,,
```
or
```csv
name,rockspec,ref,server,version,luaversion,maintainers
animation.nvim,,,,,,neorocks
```

<details><summary>Crash log of luarocks-packages-updater</summary>

```console
nixpkgs ❯ nix run nixpkgs#luarocks-packages-updater                             
can't find suitable rockspec for 'animation.nvim'

Error: No results matching query were found for Lua 5.2.
Traceback (most recent call last):
  File "/nix/store/lp65ygjnlw78hmnvz2iw4i3zl2c4wa7h-luarocks-packages-updater-0.1/bin/..luarocks-packages-updater-wrapped-wrapped", line 222, in <module>
    main()
  File "/nix/store/lp65ygjnlw78hmnvz2iw4i3zl2c4wa7h-luarocks-packages-updater-0.1/bin/..luarocks-packages-updater-wrapped-wrapped", line 218, in main
    editor.run()
  File "/nix/store/lp65ygjnlw78hmnvz2iw4i3zl2c4wa7h-luarocks-packages-updater-0.1/lib/pluginupdate.py", line 567, in run
    getattr(self, command)(args)
  File "/nix/store/lp65ygjnlw78hmnvz2iw4i3zl2c4wa7h-luarocks-packages-updater-0.1/bin/..luarocks-packages-updater-wrapped-wrapped", line 101, in update
    update_plugins(self, args)
  File "/nix/store/lp65ygjnlw78hmnvz2iw4i3zl2c4wa7h-luarocks-packages-updater-0.1/lib/pluginupdate.py", line 792, in update_plugins
    redirects = update()
                ^^^^^^^^
  File "/nix/store/lp65ygjnlw78hmnvz2iw4i3zl2c4wa7h-luarocks-packages-updater-0.1/bin/..luarocks-packages-updater-wrapped-wrapped", line 138, in update
    results = pool.map(_prefetch, sorted_plugin_specs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4/lib/python3.12/multiprocessing/pool.py", line 367, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4/lib/python3.12/multiprocessing/pool.py", line 774, in get
    raise self._value
  File "/nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4/lib/python3.12/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
                    ^^^^^^^^^^^^^^^^^^^
  File "/nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4/lib/python3.12/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
           ^^^^^^^^^^^^^^^^
  File "/nix/store/lp65ygjnlw78hmnvz2iw4i3zl2c4wa7h-luarocks-packages-updater-0.1/bin/..luarocks-packages-updater-wrapped-wrapped", line 204, in generate_pkg_nix
    output = subprocess.check_output(cmd, text=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4/lib/python3.12/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4/lib/python3.12/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['luarocks', 'nix', '--maintainers=neorocks', 'animation.nvim']' returned non-zero exit status 1.
```
</details>

Main part of the error:
```console
can't find suitable rockspec for 'animation.nvim'
Error: No results matching query were found for Lua 5.2.
```

(I tried to debug it, but I just don't understand the script: why is it trying to start a subprocess with `['luarocks', 'nix', ...]`? `luarocks` command doesn't have a `nix` subcommand.)

**Comment by teto** [comment]
ha I see. The updater uses a fork of luarocks: https://github.com/nix-community/luarocks-nix
There is no stable version, just the scm version: https://luarocks.org/modules/neorocks/animation.nvim which is why the converter doesn't find a version. Passing `--dev` works:
```
$ luarocks nix animation.nvim --dev
Error: fetchgit does not support fetching the latest revision
Call to nurl --indent 2 https://luarocks.org/dev/animation.nvim-scm-1.rockspec failed with status: 256
$ nix flake prefetch --extra-experimental-features 'nix-command flakes' --json github:anuvyklack/animation.nvim/fb77091ab72ec9971aee0562e7081182527aaa6a
Call to nurl --indent 2 https://github.com/anuvyklack/animation.nvim failed with status: 0
{ buildLuarocksPackage, fetchFromGitHub, luaOlder, middleclass, toto }:
buildLuarocksPackage {
  pname = "animation.nvim";
  version = "scm-1";
  knownRockspec = ().outPath;
  src = fetchFromGitHub {
    owner = "anuvyklack";
    repo = "animation.nvim";
    rev = "fb77091ab72ec9971aee0562e7081182527aaa6a";
    hash = "sha256-KyFzPI2B0ioIABnhC6MwaHYLcuUzSqCsmcGpMei4XXk=";
  };

  disabled = luaOlder "5.1";
  propagatedBuildInputs = [ middleclass ];

  meta = {
    homepage = "https://github.com/anuvyklack/animation.nvim";
    description = "An OOP library to create animations in Neovim.";
    license.fullName = "MIT";
  };
}
```

Not sure how to fix that properly though. Let me get back to you after work. For now you can put that definition in lua/overrides.nix to unblock you and I will fix that in follow up PRs.

**Comment by gepbird** [comment]
@teto thanks for the many useful info again, but I think I should leave this to you.

I tried to use your animation-nvim override, but `().outPath;` doesn't eval so I commented it out. Even then I get another eval error.

<details><summary>diff</summary>

```diff
diff --git a/pkgs/applications/editors/vim/plugins/overrides.nix b/pkgs/applications/editors/vim/plugins/overrides.nix
index 0a97354a14db..8e23f7341288 100644
--- a/pkgs/applications/editors/vim/plugins/overrides.nix
+++ b/pkgs/applications/editors/vim/plugins/overrides.nix
@@ -127,6 +127,8 @@
 ,
 }: self: super:
 {
+  animation-nvim-test = neovimUtils.buildNeovimPlugin { luaAttr = "animation-nvim-test"; };
+
   alpha-nvim = super.alpha-nvim.overrideAttrs {
     dependencies = [
       self.nvim-web-devicons # required by the startify theme
diff --git a/pkgs/development/lua-modules/overrides.nix b/pkgs/development/lua-modules/overrides.nix
index 99fe9217c960..bae0f5b4097e 100644
--- a/pkgs/development/lua-modules/overrides.nix
+++ b/pkgs/development/lua-modules/overrides.nix
@@ -65,6 +65,29 @@ let
   inherit (prev) luaOlder luaAtLeast lua isLuaJIT isLua51;
 in
 {
+  animation-nvim-test = final.callPackage ({ buildLuarocksPackage, fetchFromGitHub, luaOlder, middleclass, toto }:
+    buildLuarocksPackage {
+      pname = "animation.nvim-test";
+      version = "scm-1";
+      #knownRockspec = ().outPath;
+      src = fetchFromGitHub {
+        owner = "anuvyklack";
+        repo = "animation.nvim";
+        rev = "fb77091ab72ec9971aee0562e7081182527aaa6a";
+        hash = "sha256-KyFzPI2B0ioIABnhC6MwaHYLcuUzSqCsmcGpMei4XXk=";
+      };
+
+      disabled = luaOlder "5.1";
+      propagatedBuildInputs = [ middleclass ];
+
+      meta = {
+        homepage = "https://github.com/anuvyklack/animation.nvim";
+        description = "An OOP library to create animations in Neovim.";
+        license.fullName = "MIT";
+      };
+    }
+  );
+
   argparse = prev.argparse.overrideAttrs(oa: {
 
     doCheck = true;
```

</details>

<details><summary>build log</summary>

```console
nixpkgs ❯ nix build .#vimPlugins.animation-nvim-test --show-trace --no-eval-cache
warning: Git tree '/home/gep/nixpkgs' is dirty
error:
       … while calling a functor (an attribute set with a '__functor' attribute)

         at /nix/store/7ri1h10izzqm893zm7ms8xibbabb31ml-source/pkgs/applications/editors/vim/plugins/overrides.nix:130:25:

          129| {
          130|   animation-nvim-test = neovimUtils.buildNeovimPlugin { luaAttr = "animation-nvim-test"; };
             |                         ^
          131|

       … from call site

         at /nix/store/7ri1h10izzqm893zm7ms8xibbabb31ml-source/lib/trivial.nix:955:7:

          954|     { # TODO: Should we add call-time "type" checking like built in?
          955|       __functor = self: f;
             |       ^
          956|       __functionArgs = args;

       … while calling anonymous lambda

         at /nix/store/7ri1h10izzqm893zm7ms8xibbabb31ml-source/pkgs/applications/editors/neovim/build-neovim-plugin.nix:13:3:

           12|   # luaPackages
           13|   {
             |   ^
           14|     # the lua attribute name that matches this vim plugin. Both should be equal

       … from call site

         at /nix/store/7ri1h10izzqm893zm7ms8xibbabb31ml-source/pkgs/applications/editors/neovim/build-neovim-plugin.nix:32:18:

           31|
           32|       finalDrv = toVimPlugin (luaDrv.overrideAttrs(oa: attrs // {
             |                  ^
           33|           nativeBuildInputs = oa.nativeBuildInputs or [] ++ [

       … while calling 'toVimPlugin'

         at /nix/store/7ri1h10izzqm893zm7ms8xibbabb31ml-source/pkgs/applications/editors/vim/plugins/vim-utils.nix:416:17:

          415|
          416|   toVimPlugin = drv:
             |                 ^
          417|     drv.overrideAttrs(oldAttrs: {

       error: value is a function while a set was expected
```
</details>

This is getting a little out of scope of the original issue, maybe you should create a new one for generally refactoring neovim plugins made in lua.
